### PR TITLE
ctpdev:scalers and digi2raw fixes

### PR DESCRIPTION
**Pull request by lietava**
@shahor02 : New code is scalers. Plus some digi2raw fixes.

### CODE REVIEW

**Comment by lietava** [comment]
@shahor02 Is this ready for review ?

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Scalers.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  uint32_t l0After;
+  uint32_t l1Before;
+  uint32_t l1After;
+  void printStream(std::ostream& stream) const;
+  ClassDefNV(CTPScalerRaw, 1);
+};
+/// Scalers produced from raw scalers corrected for overflow
+struct CTPScalerO2 {
+  CTPScalerO2() = default;
+  void createCTPScalerO2FromRaw(CTPScalerRaw raw, std::array<uint32_t, 6>& overfow);
```
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Scalers.h]: pass `raw` by `const CTPScalerRaw& raw` and `overfow` as `const std::array<uint32_t, 6>& overfow`

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Scalers.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  // map from class index to overflow
+  // overflow counts how many time class scalerers overflowed
+  typedef std::map<uint32_t, std::array<uint32_t, 6>> overflows_t;
+  int mVersion;
+  uint32_t mRunNumber;
+  // using class mask for all class index related stuff
+  std::bitset<CTP_NCLASSES> mClassMask;
+  std::vector<CTPScalerRecordRaw> mScalerRecordRaw;
+  std::vector<CTPScalerRecordO2> mScalerRecordO2;
+  int processScalerLine(std::string& line, int& level, int& nclasses);
```
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Scalers.h]: `const std::string& line` ?

File: DataFormats/Detectors/CTP/src/Scalers.cxx [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+  for (uint32_t i = 0; i < mClassMask.size(); i++) {
+    if (mClassMask[i]) {
+      overflows[i] = {0, 0, 0, 0, 0, 0};
+    }
+  }
+  // 1st o2 rec is just copy
+  CTPScalerRecordO2 o2rec;
+  copyRawToO2ScalerRecord(mScalerRecordRaw[0], o2rec, overflows);
+  mScalerRecordO2.push_back(o2rec);
+  for (int i = 1; i < mScalerRecordRaw.size(); i++) {
```
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Scalers.cxx]: use `uint32_t` to avoid warning

File: DataFormats/Detectors/CTP/src/Scalers.cxx [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+{
+  if (rawrec.scalers.size() != (mClassMask.count())) {
+    LOG(ERROR) << "Inconsistent scaler record size:" << rawrec.scalers.size() << " Expected:" << mClassMask.count();
+    return 1;
+  }
+  o2rec.scalers.clear();
+  o2rec.intRecord = rawrec.intRecord;
+  o2rec.seconds = rawrec.seconds;
+  o2rec.microSeconds = rawrec.microSeconds;
+  for (int i = 0; i < rawrec.scalers.size(); i++) {
```
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Scalers.cxx]: idem

**Review by shahor02 — COMMENTED** [review]
Hi @lietava 
See some cosmetic comments below. Please remove the draft tag to trigger CI
Cheers, 
Ruben

**Review by shahor02 — APPROVED** [review]

**Comment by shahor02** [comment]
Hi @lietava @mbombara is the raw data decoding works correctly now? Shall I proceed to CTF creation?

**Comment by lietava** [comment]
HI @shahor02, I am not sure yet. I let you know. Thanks, Roman.

**Pull request merged**